assets/js/base/context/cart-checkout/checkout-state/index.js

```diff
@@ -0,0 +1,203 @@
+/**
+ * External dependencies
+ */
+import {
+	createContext,
+	useContext,
+	useReducer,
+	useRef,
+	useMemo,
+	useEffect,
+} from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import { actions } from './actions';
+import { reducer } from './reducer';
+import { DEFAULT_STATE, STATUS } from './constants';
+import {
+	EMIT_TYPES,
+	emitterSubscribers,
+	emitEvent,
+	emitEventWithAbort,
+	reducer as emitReducer,
+} from './event-emit';
+
+/**
+ * @typedef {import('@woocommerce/type-defs/checkout').CheckoutDispatchActions} CheckoutDispatchActions
+ * @typedef {import('@woocommerce/type-defs/contexts').CheckoutDataContext} CheckoutDataContext
+ */
+
+const CheckoutContext = createContext( {
+	submitLabel: '',
+	onSubmit: () => void null,
+	isComplete: false,
+	isIdle: false,
+	isCalculating: false,
+	isProcessing: false,
+	hasError: false,
+	redirectUrl: '',
+	orderId: 0,
+	onCheckoutCompleteSuccess: ( callback ) => void callback,
+	onCheckoutCompleteError: ( callback ) => void callback,
+	onCheckoutProcessing: ( callback ) => void callback,
+	dispatchActions: {
+		resetCheckout: () => void null,
+		setRedirectUrl: ( url ) => void url,
+		setHasError: ( hasError ) => void hasError,
+		incrementCalculating: () => void null,
+		decrementCalculating: () => void null,
+		setOrderId: ( id ) => void id,
+	},
+	hasOrder: false,
+	isCart: false,
+} );
+
+/**
+ * @return {CheckoutDataContext} Returns the checkout data context value
+ */
+export const useCheckoutContext = () => {
+	return useContext( CheckoutContext );
+};
+
+/**
+ * Checkout state provider
+ * This provides provides an api interface exposing checkout state for use with
+ * cart or checkout blocks.
+ *
+ * @param {Object}  props                     Incoming props for the provider.
+ * @param {Object}  props.children            The children being wrapped.
+ * @param {string}  props.redirectUrl         Initialize what the checkout will
+ *                                            redirect to after successful
+ *                                            submit.
+ * @param {string}  props.submitLabel         What will be used for the checkout
+ *                                            submit button label.
+ * @param {boolean} props.isCart              If context provider is being used
```

Comment:
How does `isCart` get used inside the provider/context? This feels a little suspect, maybe just the name. If it's to disable certain things, can we expose options for disabling them?

Reply:
> How does isCart get used inside the provider/context? 

`CartProvider` and `CheckoutProvider` are components that implement this particular provider along with any other providers they need to use. The challenge here is that children components within the providers _may_ need to know what context they are in (cart or checkout) so this property provides an easy way to expose that context. There are no implementations of this flag yet but one implementation where I could see this being useful is for customizing css classes on shared components.

>  If it's to disable certain things, can we expose options for disabling them?

Sorry I don't know what you mean by "expose options for disabling them".

Reply:
Ah so to be clear, this is not currently used?

Reply:
No.

Reply:
I see! Usually I'd recommend not adding this until it's needed / used. No biggie since is already merged.

Reply:
Just to followup, there actually [is a place where this can be implemented now](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2101#discussion_r402440162).  _I thought I had seen this kind of flag in use somewhere_ :)